### PR TITLE
Add Knox init action 

### PR DESCRIPTION
**Pull request by cyxxy**
continuing #604

### CODE REVIEW

File: knox/topologies/example-hive-nonpii.xml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0"?>
```
Comment (medb) [inline knox/topologies/example-hive-nonpii.xml]: @yunus How this file is used?

File: knox/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+### Caveats
+
+ - Dataproc Jobs API uses binary transport protocol of Hive. When we switch the protocol to HTTP for Knox, you cannot use the Jobs API for these backend clusters. See [Additional setup for Hive](#additional-setup-for-hive) for details.
+ - [Component Gateway](https://cloud.google.com/dataproc/docs/concepts/accessing/dataproc-gateways) does not work for the gateway cluster when *SSL* is enabled by changing [gateway-site.xml](./gateway-site.xml) or when [default.xml](./topologies/default.xml) is altered since component gateway also uses Knox. Note that, the backend clusters that gateway redirects can still use the component gateway.
+
+## Using this initialization action
+
+This initialization action requires a bucket that stores configuration. By providing a bucket with configurations, you can create a Dataproc cluster with Knox installed as follows:
+
```
Comment (medb) [inline knox/README.md]: Add disclaimer/notice about using init actions in prod (see other init action docs)

File: knox/README.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+1.  Use the `gcloud` command to create a new cluster with this initialization
+    action. 
+
+    ```bash
+    gcloud dataproc clusters create <CLUSTER_NAME> \
```
Comment (medb) [inline knox/README.md]: Modify command to use regional GCS buckets by default.

File: knox/knox.sh [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		       -keystore $GATEWAY_HOME/data/security/keystores/gateway.jks -storepass $master_key -noprompt
+		sudo -u knox keytool -importcert -file $GATEWAY_HOME/data/security/keystores/gateway-client.crt \
+		       -keystore $GATEWAY_HOME/data/security/keystores/gateway-client.jks -alias "gateway-identity" -storepass $master_key -noprompt
+		
+		sudo -u knox $GATEWAY_HOME/bin/knoxcli.sh export-cert --type PEM
+
+		gsutil cp $GATEWAY_HOME/data/security/keystores/gateway-identity.pem gs://$KNOX_GW_CONFIG_GS/exported-certs/
+		gsutil cp $GATEWAY_HOME/data/security/keystores/gateway-client.jks gs://$KNOX_GW_CONFIG_GS/exported-certs/
+
+	elif [[ ! -z "${custom_cert_name}" ]]; then
```
Comment (medb) [inline knox/knox.sh]: Replace `! -z` with `-n`

File: knox/knox.sh [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+# limitations under the License.
+
+# This init script installs Apache Zeppelin on the master node of a Cloud
+# Dataproc cluster. Zeppelin is also configured based on the size of your
+# cluster and the versions of Spark/Hadoop which are installed.
+
+set -euxo pipefail
+
+readonly KNOX_GW_CONFIG_GS=$(/usr/share/google/get_metadata_value attributes/knox-gw-config-gs)
+readonly KNOX_GW_CONFIG=/knox-config
```
Comment (medb) [inline knox/knox.sh]: create temp work directory instead of hard-coding root directory.

File: knox/knox.sh [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This init script installs Apache Zeppelin on the master node of a Cloud
+# Dataproc cluster. Zeppelin is also configured based on the size of your
+# cluster and the versions of Spark/Hadoop which are installed.
+
+set -euxo pipefail
+
+readonly KNOX_GW_CONFIG_GS=$(/usr/share/google/get_metadata_value attributes/knox-gw-config-gs)
```
Comment (medb) [inline knox/knox.sh]: I think we can remove `-gs` suffix from the metadata property name - it seems redundant.

Comment (cyxxy) [reply knox/knox.sh]: done.

File: knox/knox.sh (hunk 5 above)
Comment (cyxxy) [reply knox/knox.sh]: done.

File: knox/knox.sh (hunk 4 above)
Comment (cyxxy) [reply knox/knox.sh]: done.

File: knox/README.md (hunk 3 above)
Comment (cyxxy) [reply knox/README.md]: done.

File: knox/README.md (hunk 2 above)
Comment (cyxxy) [reply knox/README.md]: done.

File: knox/README.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+```
+
+For parallel runs, you may use [python3-fastunit](https://github.com/ityoung/python3-fastunit). For details checkout [integration tests](../integration_tests).
+
+## Additional setup for Hive
+
+To use Knox as a Hive gateway, you should configure Hive and set the transport protocol to Http. Knox currently does not support the Hive binary transport protocol. The below command sets the required properties:
+
+```bash
+    gcloud dataproc clusters create <CLUSTER_NAME> \
```
Comment (medb) [inline knox/README.md]: Change to use regional buckets

File: knox/README.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+1.  Use the `gcloud` command to create a new cluster with this initialization
+    action. 
+
+    ```bash
+    REGION=<region>
+    CLUSTER_NAME=<cluster_name>
+    gcloud dataproc clusters create ${CLUSTER_NAME} \
+        --region ${REGION}
+        --initialization-actions gs://dataproc-initialization-actions-${REGION}/knox/knox.sh \
+        --metadata knox-gw-config=<your knox configuration directory without gs:// prefix>
```
[the remark this answers is not in the document]
Comment (cyxxy) [reply knox/README.md]: cannot be the regional initialization-actions bucket. It has to be owned by the user.

File: knox/README.md (hunk 7 above)
Comment (cyxxy) [reply knox/README.md]: done.

File: knox/README.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+## Automated tests
+
+This init action can be tested with automated script `test_knox.py`. In order to
+run just use the below command from project top directory.
+
+```bash
+python3 -m unittest knox.test_knox -v
+```
+
+For parallel runs, you may use [python3-fastunit](https://github.com/ityoung/python3-fastunit). For details checkout [integration tests](../integration_tests).
```
[the remark this answers is not in the document]
Comment (cyxxy) [reply knox/README.md]: done

File: knox/README.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+3.  To learn about how to use Knox read the [user guide](https://knox.apache.org/books/knox-1-1-0/user-guide.html).
+
+## Automated tests
+
+This init action can be tested with automated script `test_knox.py`. In order to
+run just use the below command from project top directory.
+
+```bash
+python3 -m unittest knox.test_knox -v
```
[the remark this answers is not in the document]
Comment (cyxxy) [reply knox/README.md]: done.

File: knox/README.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    REGION=<region>
+    CLUSTER_NAME=<cluster_name>
+    gcloud dataproc clusters create ${CLUSTER_NAME} \
+        --region ${REGION}
+        --initialization-actions gs://dataproc-initialization-actions-${REGION}/knox/knox.sh \
+        --metadata knox-gw-config=<your knox configuration directory without gs:// prefix>
+    ```
+    The configuration bucket should be a modified copy of this git directory. The bucket structure is:
+    ```
+     <your gs bucket>
```
[the remark this answers is not in the document]
Comment (cyxxy) [reply knox/README.md]: done

File: knox/topologies/example-hive-nonpii.xml (hunk 1 above)
Comment (yunus) [reply knox/topologies/example-hive-nonpii.xml]: It is an example configuration where a guest account is allowed to access wedhdfs. I have used this in tests.

File: knox/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (yunus) [reply knox/README.md]: that's true, user should own the configuration bucket. It is the place where all the knox config and topology files exist.

**Comment by yunus** [comment]
@cyxxy I am not able to test against version 1.5, but for 1.3 and 1.4, tests pass in my local environment. Tell me if you need help/explanation.

**Pull request merged**